### PR TITLE
server: call server.address after listening event

### PR DESCRIPTION
**Pull request by kjdelisle**
As per the Node.js API docs (https://nodejs.org/api/http.html)
the `server.address` function should only be called *after*
the "listening" event is triggered.

cc @bajtos @raymondfeng @ritch @superkhau

### CODE REVIEW

**Comment by kjdelisle** [comment]
By the way, is the "cc" a PR template? I don't remember typing that. :P

**Comment by kjdelisle** [comment]
I strongly suspect some of the current test logic is assuming it can proceed before the listening event (and consequently, the `server.start`call) have returned.

**Comment by superkhau** [comment]
Can you fix the tests before landing? The changes LGTM.

**Comment by raymondfeng** [comment]
@kjdelisle The CI is failing. Can you fix it?

**Comment by kjdelisle** [comment]
@superkhau @raymondfeng Sorry that took so long, was fighting with an unrelated task.
I think that my changes are floating on some sort of middle ground where Coveralls randomly decides that I have, or haven't decreased code coverage...
It keeps flipping between green and red.

File: packages/core/src/server.ts [hunk 1]
```diff
@@ -20,19 +20,27 @@ export class Server extends Context {
...
     server.listen(this.config.port);
 
-    // FIXME(bajtos) The updated port number should be part of "status" object,
-    // we shouldn't be changing original config IMO.
-    // Consider exposing full base URL including http/https scheme prefix
-    this.config.port = server.address().port;
 
-    await new Promise(resolve => server.once('listening', resolve));
-    this.state = ServerState.listening;
+    return new Promise<void>((resolve, reject) => {
```
Comment (bajtos) [inline packages/core/src/server.ts]: @kjdelisle it makes me wonder, is there any particular reason against using `await` here?

```ts
await new Promise(resolve => server.once('listening', resolve));
this.config.port = server.address().port;
this.state = ServerState.listening;
```

File: packages/core/test/acceptance/routing/routing.acceptance.ts [hunk 2]
```diff
@@ -61,12 +61,11 @@ describe('Routing', () => {
     }
     givenControllerInApp(app, EchoController);
 
-    return (
-      whenIMakeRequestTo(app)
-        .get('/echo?msg=hello%20world')
-        // Then I get the result `hello world` from the `Method`
-        .expect('hello world')
-    );
+    return whenIMakeRequestTo(app).then(client => {
```
Comment (bajtos) [inline packages/core/test/acceptance/routing/routing.acceptance.ts]:
```ts
const client = await whenIMakeRequestTo(app);
return client.get(...)
```

**Pull request merged**